File: homework6/src/main/java/by/popkov/homework6/EditContactActivity.java
package by.popkov.homework6;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.InputType;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;


public class EditContactActivity extends AppCompatActivity {
    public static final String EXTRA_NEW_CONTACT = "newContact";
    public static final String EXTRA_OLD_CONTACT = "contact";
    public static final String EXTRA_FULL_LIST_POS = "fullListPosition";
    public static final String EXTRA_LIST_POS = "listPosition";

    private EditText editTextName;
    private EditText editTextPhoneNumberOrEmail;
    private Button buttonEdit;
    private Button buttonRemove;

    private Contact oldContact;
    private int fullListPosition;
    private int listPosition;


    public static Intent newIntent(Context context, Contact contact, int fullListPosition, int listPosition) {
        Intent intent = new Intent(context, EditContactActivity.class);
        intent.putExtra(EXTRA_OLD_CONTACT, contact);
        intent.putExtra(EXTRA_FULL_LIST_POS, fullListPosition);
        intent.putExtra(EXTRA_LIST_POS, listPosition);
        return intent;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_edit_contact);
        init();
        setDataInFields();
        setListeners();
        setToolBar();
    }

    private void setToolBar() {
        setSupportActionBar((Toolbar) findViewById(R.id.toolBar));
        ActionBar supportActionBar = getSupportActionBar();
        if (supportActionBar != null) {
            supportActionBar.setDisplayHomeAsUpEnabled(true);
            supportActionBar.setDisplayShowHomeEnabled(true);
        }
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void setListeners() {
        buttonRemove.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(EditContactActivity.this, ContactsActivity.class);
                intent.putExtra(EXTRA_FULL_LIST_POS, fullListPosition);
                intent.putExtra(EXTRA_LIST_POS, listPosition);
                setResult(RESULT_OK, intent);
                finish();
            }
        });
        buttonEdit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(EditContactActivity.this, ContactsActivity.class);
                intent.putExtra(EXTRA_FULL_LIST_POS, fullListPosition);
                intent.putExtra(EXTRA_LIST_POS, listPosition);
                String name = editTextName.getText().toString().trim();
                String data = editTextPhoneNumberOrEmail.getText().toString().trim();
                Contact newContact;
                if (oldContact == Contact.EMAIL) {
                    newContact = Contact.EMAIL
                            .setName(name)
                            .setData(data);
                } else {
                    newContact = Contact.PHONE
                            .setName(name)
                            .setData(data);
                }
                intent.putExtra(EXTRA_NEW_CONTACT, newContact);
                setResult(RESULT_OK, intent);
                finish();
            }
        });
    }

    private void setDataInFields() {
        if (oldContact != null) {
            if (oldContact == Contact.EMAIL) {
                editTextPhoneNumberOrEmail.setHint(R.string.email);
                editTextPhoneNumberOrEmail.setInputType(InputType.TYPE_TEXT_VARIATION_EMAIL_ADDRESS);
            } else {
                editTextPhoneNumberOrEmail.setHint(R.string.phone_number);
                editTextPhoneNumberOrEmail.setInputType(InputType.TYPE_CLASS_PHONE);
            }
            editTextName.setText(oldContact.getName());
            editTextPhoneNumberOrEmail.setText(oldContact.getData());
        }


    }

    private void init() {
        editTextName = findViewById(R.id.editTextName);
        editTextPhoneNumberOrEmail = findViewById(R.id.editTextPhoneNumberOrEmail);
        buttonEdit = findViewById(R.id.buttonEdit);
        buttonRemove = findViewById(R.id.buttonRemove);
        Intent comeIntent = getIntent();
        oldContact = (Contact) comeIntent.getSerializableExtra(EXTRA_OLD_CONTACT);
        fullListPosition = comeIntent.getIntExtra(EXTRA_FULL_LIST_POS, -404);
        listPosition = comeIntent.getIntExtra(EXTRA_LIST_POS, -777);
    }
}

File: homework4/src/main/java/by/popkov/homework4/EditContactActivity.kt
package by.popkov.homework4

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.InputType
import android.view.MenuItem
import android.widget.Button
import android.widget.EditText
import androidx.appcompat.app.AppCompatActivity


class EditContactActivity : AppCompatActivity() {
    private lateinit var editTextName: EditText
    private lateinit var editTextPhoneNumberOrEmail: EditText
    private lateinit var buttonEdit: Button
    private lateinit var buttonRemove: Button

    private lateinit var oldContact: Contact
    private var fullListPosition: Int = -100
    private var listPosition: Int = -100

    companion object {
        const val EXTRA_NEW_CONTACT = "newContact"
        const val EXTRA_OLD_CONTACT = "contact"
        const val EXTRA_FULL_LIST_POS = "fullListPosition"
        const val EXTRA_LIST_POS = "listPosition"

        fun newIntent(context: Context, contact: Contact, fullListPosition: Int, listPosition: Int): Intent {
            val intent = Intent(context, EditContactActivity::class.java)
            intent.putExtra(EXTRA_OLD_CONTACT, contact)
            intent.putExtra(EXTRA_FULL_LIST_POS, fullListPosition)
            intent.putExtra(EXTRA_LIST_POS, listPosition)
            return intent
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_edit_contact)
        init()
        setDataInFields()
        setListeners()
        setToolBar()
    }

    private fun setToolBar() {
        setSupportActionBar(findViewById(R.id.toolBar))
        val supportActionBar = supportActionBar
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setDisplayShowHomeEnabled(true)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun setListeners() {
        buttonRemove.setOnClickListener {
            val intent = Intent(this@EditContactActivity, ContactsActivity::class.java)
            intent.putExtra(EXTRA_FULL_LIST_POS, fullListPosition)
            intent.putExtra(EXTRA_LIST_POS, listPosition)
            setResult(Activity.RESULT_OK, intent)
            finish()
        }
        buttonEdit.setOnClickListener {
            val intent = Intent(this@EditContactActivity, ContactsActivity::class.java)
            intent.putExtra(EXTRA_FULL_LIST_POS, fullListPosition)
            intent.putExtra(EXTRA_LIST_POS, listPosition)
            val name = editTextName.text.toString().trim()
            val data = editTextPhoneNumberOrEmail.text.toString().trim()
            val newContact: Contact
            newContact = if (oldContact is ContactEmail) ContactEmail(name, data, Contact.IMAGE_ID_EMAIL)
            else ContactPhone(name, data, Contact.IMAGE_ID_PHONE)
            intent.putExtra(EXTRA_NEW_CONTACT, newContact)
            setResult(Activity.RESULT_OK, intent)
            finish()
        }
    }

    private fun setDataInFields() {
        if (oldContact is ContactEmail) {
            editTextPhoneNumberOrEmail.setHint(R.string.email)
            editTextPhoneNumberOrEmail.inputType = InputType.TYPE_TEXT_VARIATION_EMAIL_ADDRESS
        } else {
            editTextPhoneNumberOrEmail.setHint(R.string.phone_number)
            editTextPhoneNumberOrEmail.inputType = InputType.TYPE_CLASS_PHONE
        }
        editTextName.setText(oldContact.name)
        editTextPhoneNumberOrEmail.setText(oldContact.data)
    }

    private fun init() {
        editTextName = findViewById(R.id.editTextName)
        editTextPhoneNumberOrEmail = findViewById(R.id.editTextPhoneNumberOrEmail)
        buttonEdit = findViewById(R.id.buttonEdit)
        buttonRemove = findViewById(R.id.buttonRemove)
        val comeIntent = intent
        oldContact = comeIntent.getSerializableExtra(EXTRA_OLD_CONTACT) as Contact
        fullListPosition = comeIntent.getIntExtra(EXTRA_FULL_LIST_POS, -404)
        listPosition = comeIntent.getIntExtra(EXTRA_LIST_POS, -777)
    }
}